### PR TITLE
chore(CONTRIBUTING.md): enforce commit guidelines

### PR DESCRIPTION
**Pull request by thunder-coding**
Requesting reviews from @Grimler91 @xtkoba @buttaface @MrAdityaAlok.

Feel free to ping any other organization member who'll be affected by this change

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+In order to encourage new contributors and help them conteibute to open source, the above mentioned commit requirements should be optionally relaxed. In cases where commit messages need to be changed, the PR may be **Squashed and Merged** or may be merged manually from the command line.
+
+#### Notes for merging PRs from command line
+
+1. It is recommended to use the [GitHub CLI (`gh`)](https://cli.github.com) in order to fetch the contributor's branch.
+   ```sh
+   gh pr checkout <PR Number>
+   ```
+
+2. After checking out the branch, ammend the commit message and optionally rebase against the master branch (if necessary).
```
Comment (Grimler91) [inline CONTRIBUTING.md:355]: typo, ammend -> amend

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+
+   When merging manually make sure that you give proper credits for the original patch to it's author by adding a `Co-authored-by: ` line. See https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors for more details. Also add a `Closes #<PR number>`
```
Comment (Grimler91) [inline CONTRIBUTING.md:357]: Isn't something like

```
git rebase .....
git push -f the_fork HEAD:fork_branch
git push origin master
```

better than `Co-authored-by:` and `Closes:` as it also marks the PR as merged in github's web ui? (probably possible with `gh` as well if fork branch can be force-updated before merge)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+
+   ```sh
+   git commit --amend # Will open up your editor to ammend the commit message
```
Comment (Grimler91) [inline CONTRIBUTING.md:360]: ammend -> amend

File: CONTRIBUTING.md (hunk 2 above)
Comment (Grimler91) [reply CONTRIBUTING.md:357]: Can also considering forking and promoting something like [postmarketos's merge request helper](https://gitlab.com/postmarketOS/mrhlpr). I have a local fork of it that works with github, but maybe it is easier to add a small bash script to do the same operations with `git` or `gh`.

File: CONTRIBUTING.md (hunk 3 above)
Comment (Grimler91) [reply CONTRIBUTING.md:360]: Makes more sense to have `git commit --amend` after `git rebase origin/master`. If the amended commit is to be signed (`git commit --amend -S`) it otherwise have to be signed twice

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+3. Note down the branch name
+   ```sh
+   git branch
+   ```
+
+4. Merge the branch manually
+   ```sh
+   git switch master
+
+   git merge <branch name>
```
Comment (Grimler91) [inline CONTRIBUTING.md:375]: Maybe add a note that the merge branched is fast-forwarded by default (so no merge commit is created, unlike when PRs are merged through github's web ui)

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+   ```
+
+5. Congratulate the user on sending their (probably) first OSS contribution!
```
Comment (Grimler91) [inline CONTRIBUTING.md:378]: :tada:

**Review by Grimler91 — COMMENTED** [review]
Looks mostly good to me, thanks for formalising this!

We can add a lint script that checks the commit message style as well, so that user has a chance to fix style issues without maintainers nagging about it

**Comment by Grimler91** [comment]
> I don't like this.

Writing guidelines on how commits should preferably be written should be a good idea in any case.
Is your preference the format `package: some description`?

**Comment by xtkoba** [comment]
> Writing guidelines on how commits should preferably be written should be a good idea in any case.

This is true. We have seen a bunch of PRs of which the subject is not comprehensive.

Yes I prefer `package: Some description` and `new package: pkgname`, because I believe they are comprehensive and easy to write.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+    Optional (but highly recommended) extended commit message may contain detailed summary of the enabled feature(s) and a basic use case
+  - `chore`: Any housekeeping change or a change which does not affect the user in any way.
+  - `rebuild`: Rebuild a package to link against newer version of shared library
+    Special cases:
+    - When mass rebuilding packages depending on a major package (e.g. openssl), consider using this format:
+      ```
+      rebuild(deps:main/openssl): link against OpenSSL 3.0
+      ```
+  - `scripts`: Any change affecting our build scripts
+
```
Comment (MrAdityaAlok) [inline CONTRIBUTING.md:326]: Also add `disable(<repo>/<pkg name>): disable a package due to some issue`. Reason to disable should be mandatory.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+3. ```
+   enhance,upgpkg(nodejs): update to v18.2.0 and use shared libuv
+
+   # Describe the technical reasons of how using shared libuv is beneficial
+   ```
+
+### Special notes for newbies who're just getting started with Open Source
+
+In order to encourage new contributors and help them conteibute to open source, the above mentioned commit requirements should be optionally relaxed. In cases where commit messages need to be changed, the PR may be **Squashed and Merged** or may be merged manually from the command line.
```
Comment (MrAdityaAlok) [inline CONTRIBUTING.md:346]: Typo: `conteibute`

**Review by MrAdityaAlok — CHANGES_REQUESTED** [review]

**Comment by xtkoba** [comment]
My proposal:

 * If a commit is about to modify a single package, then the subject should contain the package name.

And only this. Unnecessarily complex rules are just hard to obey.

**Comment by truboxl** [comment]
How long should commit message be?
I still prefer xtkoba's way as well since the proposal add quite a bit characters and not fit 50 chars for long package names...

**Comment by agnostic-apollo** [comment]
The suggested rules by @thunder-coding look good to me too. They are following the conventional commit design, which is used by lot of FOSS projects now and also the termux apps repos mostly. Deciding scope as package name is easier here than app components though.

https://www.conventionalcommits.org/en/v1.0.0

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+## Commit guidelines
+
+A commit message should look something like:
+```
+<commitType>(<repo>/<package>): (The change made/Short description of the change)
+
+[An optional but **highly recommended** commit message describing the changes made in the commit]
+
+[Fixes https://github.com/termux/termux-packages/issues/<issue number>]
+[Closes https://github.com/termux/termux-packages/pull_requests/<pr number>]
```
[the remark this answers is not in the document]
Comment (agnostic-apollo) [reply CONTRIBUTING.md:308]: Or `termux/some-repo#<number>` for other repos.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -296,6 +296,89 @@ OS and is built by Android NDK. This often introduces compatibility issues as
 Android (specifically Termux) is not a standard platform. Do not expect there
 are exist package recipes available out-of-box.
 
+## Commit guidelines
+
+A commit message should look something like:
+```
+<commitType>(<repo>/<package>): (The change made/Short description of the change)
```
[the remark this answers is not in the document]
Comment (thunder-coding) [reply CONTRIBUTING.md:303]: Choosing a length of 80, used by many projects. Also most editor formatting defaults make use of 80 as line length

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -296,6 +296,113 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
+
+[An optional but **highly recommended** commit message describing the changes made in the commit]
+
+[Fixes (termux/repo)#<issue number>]
+[Closes (termux/repo)#<pr number>]
+```
+
+Any line in the commit **should not exceed 80 characters**. In case it does, consider using different wordings or language style which better summarizes the changes done.
+
+- `<commitType>` describes the type of commit. Types of commits:
```
Comment (Grimler91) [inline CONTRIBUTING.md:313]: What about bug fixes (as in https://github.com/termux/termux-packages/commit/38bd77272ade88216ca1b6c2dde9b5e44feda96e just now for example)? `enhance` doesn't quite seem to describe that case

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -296,6 +296,113 @@ OS and is built by Android NDK. This often introduces compatibility issues as
...
 
+## Commit guidelines
+
+A commit message should look something like:
+```
+<commitType>(<repo>/<package>): (The change made/Short description of the change)
+
+[An optional but **highly recommended** commit message describing the changes made in the commit]
+
+[Fixes (termux/repo)#<issue number>]
```
Comment (Grimler91) [inline CONTRIBUTING.md:307]: is this pattern "recognised" by github's web UI? 

Tests:

Just pure link: https://github.com/termux/termux-packages/issues/7834
With parentheses: (termux/termux-packages)#7834

just providing a link like https://github.com/termux/termux-packages/pull/10748 is quite handy

Comment (Grimler91) [reply CONTRIBUTING.md:307]: Without parentheses works, so I suggest we use that

File: CONTRIBUTING.md (hunk 10 above)
Comment (MrAdityaAlok) [reply CONTRIBUTING.md:313]: I suggest to use `fix(scope): msg`. Scope can be anything already defined in guideline.

File: CONTRIBUTING.md (hunk 11 above)
Comment (thunder-coding) [reply CONTRIBUTING.md:307]: Well, parentheses was intended to show that it is optional. That is, when referencing issue/PR from the same repository, the details inside parentheses would not be needed. But when referencing issues from other repositories, it's needed. It seems like this is misleading.

What do you think of something like:
```
Fixes #<issue number>
Fixes termux/<repo>#<issue number>
```

I guess this will not be mistakened

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -296,6 +296,113 @@ OS and is built by Android NDK. This often introduces compatibility issues as
 Android (specifically Termux) is not a standard platform. Do not expect there
 are exist package recipes available out-of-box.
 
+## Commit guidelines
+
+A commit message should look something like:
+```
+<commitType>(<repo>/<package>): (The change made/Short description of the change)
```
Comment (truboxl) [inline CONTRIBUTING.md:303]: Is this repo or branch? I see that "main" is used occasionally...

File: CONTRIBUTING.md (hunk 2 above)
Comment (MrAdityaAlok) [reply CONTRIBUTING.md]: Also, github doesn't allow editing (or force pushing) PRs from org accounts.
I came across this today in #10894

File: CONTRIBUTING.md (hunk 12 above)
Comment (Grimler91) [reply CONTRIBUTING.md:303]: repo/name, main, root or x11 (which is what they are called in [repo.json](https://github.com/termux/termux-packages/blob/master/repo.json)), but yes, could be clarified

Comment (thunder-coding) [reply CONTRIBUTING.md]: Thanks, mentioned the same in the guidelines as well in order to avoid confusion.

Hence resolving

File: CONTRIBUTING.md (hunk 10 above)
Comment (thunder-coding) [reply CONTRIBUTING.md:318]: A 'fix(scope): message' rule has been added as suggested by @MrAdityaAlok.

Hence resolving

**Comment by 2096779623** [comment]
@thunder-coding https://github.com/termux/termux-packages/pull/11281

**Comment by thunder-coding** [comment]
Lol it's the typo I made earlier, which I fixed somewhere between. It's `upgpkg` only standing for **upg**rade **p**ac**k**a**g**e.

Anyways it's inspired by Arch's commit messages used in their VCS.

**Comment by thunder-coding** [comment]
About adding scripts to lint commit messages, I don't think that's the priority for me now. In case anyone is interested in it, I would be very glad. Atleast I don't expect me to work on it. In case everything seems good, and I get a approving review, I'll format the markdown file, (possibly affecting other lines than the changed ones too), and merge it.

**Comment by thunder-coding** [comment]
Let's finally get this merged. I see a lot do new PRs following similar guidelines, so making it official from semioffical might be worth it :)

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -298,6 +304,154 @@ OS and is built by Android NDK. This often introduces compatibility issues as
 Android (specifically Termux) is not a standard platform. Do not expect there
 are exist package recipes available out-of-box.
 
+## Commit guidelines
+
+A commit message should look something like:
```
Comment (Grimler91) [inline CONTRIBUTING.md:309]: Most important should be that the commit messages are logical, i.e. states what is done and what packages are affected. 

Since not all contributors agree that this style is preferable I suggest we make the style preferable but optional. We can add an introduction paragraph here like

```
Commit messages should describe the changes done, so that maintainers can understand what was done, and to what package or scope, without having to look at the code changes. One good (but not mandatory) way to make sure the commit message fulfills these requirements is to write it on the format:
```

Comment (agnostic-apollo) [reply CONTRIBUTING.md:309]: To elaborate on that. A commit message should normally explain what is being changed, why it is being changed or the effect of the commit and optionally (summary of) how changes were made if the changes were logically complex. Whoever may be reading the commit messages may not understand the code itself, like the author does or at all.

Maybe a link to some resource site would be useful. Maybe something like https://who-t.blogspot.com/2009/12/on-commit-messages.html or https://wiki.openstack.org/wiki/GitCommitMessages

>Since not all contributors agree that this style is preferable I suggest we make the style preferable but optional. 

Yeah, I guess that's reasonable. But would be better to restrict it to either `package: Some description` or `change_type(package): Some description` formats, since that way there will at least be some consistency.

**Comment by agnostic-apollo** [comment]
>adapt commit messages to new guiselines

 What's that? Second typo ;)

File: CONTRIBUTING.md (hunk 13 above)
Comment (thunder-coding) [reply CONTRIBUTING.md:309]: > Yeah, I guess that's reasonable. But would be better to restrict it to either package: Some description or change_type(package): Some description formats, since that way there will at least be some consistency.

I totally agree on this. Although it might not seem good forcefully do so, but definitely helps when going through old commits.

**Comment by thunder-coding** [comment]
> > adapt commit messages to new guiselines
> 
> What's that? Second typo ;)

So you've been spying on the commit log where you saw the old typo fix :) Anyways fixed. Thanks\

File: CONTRIBUTING.md (hunk 13 above)
Comment (agnostic-apollo) [reply CONTRIBUTING.md:309]: We are not forcing everyone to use specific formats for all repos, just our own termux-packages. Open source projects often have specific formats for their own repos, which need to be followed by all internal and external devs contributing to it. People can use whatever they want in their own repos. Currently, internal devs don't agree on one format and we have 2 different formats being used, which are both fine and will keep history clean, so might as well just restrict it to those 2 for all internal and external devs, instead of allowing commiters to use random formats and messing with the history.

Comment (thunder-coding) [reply CONTRIBUTING.md]: Done. Thanks, for all the valuable feedback. Not forcing to follow these new guidelines might be helpful as I agree that these guidelines are pretty hard to follow. I hope we may soon in future get used to these, though :)

**Comment by thunder-coding** [comment]
I'll merge this PR tommorow if there are no changes requested by anyone else.
